HuffmanCoding/HuffDecode.cpp

```diff
@@ -6,38 +6,128 @@
 #include <string>
 #include <algorithm>
 #include <map>
+#include <sstream>
+
 
 
 HuffDecode::HuffDecode() {
-	
+	bitcount = 8;
+}
+
+HuffDecode::~HuffDecode() {
 	
 }
 
 
+
 void HuffDecode::decode() {
+	std::ofstream f;
+	std::string str;
+	std::string huf;
+
+
+	
+	for (int i = 0; i < text.length(); i++) {
+
+		str.append(sizeof(char),text.at(i));
+		char c = findChar(str);
 
+		if (c) {
+			huf.append(sizeof(char), c);
+			str.clear();
+		}
+		
+
+
+	}
+	f.open("decoded.txt");
+	f << huf;
+	f.close();
 }
 
+
+
 //use stored_map to decode the encoded file 
+void HuffDecode:: readEncodedFile() {
+	std::ifstream file;
+	std::ofstream out_file;
 
+	std::string bitfreq;
+	int ch;
+	file.open("The Gunslinger.huf", std::ios::in | std::ios::binary); //read from binary file and write again txt file.
+	file.read(&acc, sizeof(acc));
+	
+	while (!file.eof()) {
+		int bit = readBit(file);
+		char ch = '0' + bit;
+		text.append(sizeof(char),ch);
+	
+		
+	}
+	file.close();
 
+}
 
+	
+auto HuffDecode:: returnMap() {
+	return newmap;
+}
 
-int HuffDecode::readBit()
+
+int HuffDecode::readBit(std::ifstream &fname)
 {
 	if (bitcount == 0)
 	{
 		bitcount = 8;
-		///acc = readbyte();            
+		//acc = readEncodedFile(newmap); //this function must be non-void.
+		fname.read(&acc, sizeof(acc));
+		
 	}
 	--bitcount;
 	return (acc >> (7 - bitcount)) & 1;
 }
 
 void HuffDecode::readDictFile() {
 	std::ifstream file;
-
+	char ch;
+	std::string bitnum;
+	std::string line;
 	file.open("The Gunslinger.dict");
-	
+	/*std::stringstream ss(line);*/
+	if (file.is_open()) {
+		while (!file.eof()) {
+			getline(file, line);
+			if (line.empty())
+				continue;
+			if (line[0] == ' ' && line[1] != ' ') {
+				newmap['\n'] = line.substr(1, line.size());
```

Comment:
Read carefully the documentation of `std::string::substr`. Here, it works because it is seeing a null terminator before going further. However, your second parameter is not correct.